app_history.md

```diff
@@ -0,0 +1,423 @@
+  document.body.querySelector("#homepage").hidden = false;
+});
+appHistory.currentEntry.addEventListener("navigatefrom", () => {
+  document.body.querySelector("#homepage").hidden = true;
+});
+</script>
+```
+
+These events are also cancelable; if they are canceled, using `event.preventDefault()`, then attempts to navigate to the app history entry in question will fail:
+
+```js
+appHistory.currentEntry.addEventListener("navigatefrom", e => {
+  if (hasUnsavedData()) {
+    displayMessage("Are you sure you want to leave? You have unsaved data.");
+    e.preventDefault();
+  }
+});
+```
+
+_TODO: maybe this would be better with `respondWith()`? Then the promise could fulfill/reject..._
+
+Additionally, there's a `dispose` event, which occurs when an app history entry is permanently evicted and unreachable: for example, in the following scenario.
+
+```js
+const startingKey = appHistory.currentEntry.key;
+
+appHistory.pushNewState();
+appHistory.currentEntry.addEventListener("dispose", () => console.log(1));
+
+appHistory.pushNewState();
+appHistory.currentEntry.addEventListener("dispose", () => console.log(2));
+
+appHistory.pushNewState();
+appHistory.currentEntry.addEventListener("dispose", () => console.log(3));
+
+appHistory.navigateTo(startingKey);
+appHistory.pushNewState();
+
+// Logs 1, 2, 3 as that branch of the tree gets pruned.
+```
+
+This can be useful for cleaning up any information in secondary stores, such as `sessionStorage` or caches, when we're guaranteed to never reach those particular history entries again.
+
+_TODO: it seems like the event probably needs to include the disposed app history entry? Maybe all of these events should contain the history entry?_
+
+#### Current entry change monitoring
+
+The `window.appHistory` object has an event, `currententrychange`, which allows the application to react to any updates to the `appHistory.currentEntry` property. This includes both navigations that change its value, and calls to `appHistory.updateCurrentEntry()` that change its `state` or `url` properties.
+
+This cannot be intercepted or canceled, as it occurs after the navigation has already happened; it's just an after-the-fact notification. An example of how an application might use it is as follows:
+
+```js
+appHistory.addEventListener("currententrychange", () => {
+  document.querySelector("#order-status") = appHistory.currentEntry.state.?orderStatus ?? "";
+});
+```
+
+#### Navigation monitoring and interception
+
+Finally, the most interesting event on `window.appHistory` is the one which allows monitoring and interception of navigations: the `navigate` event. It fires on any navigation, either user-initiated or application-initiated, which would update the value of `appHistory.currentEntry`. This includes cross-origin navigations (which will take us out of the current app history list), but it does _not_ include navigations which target other windows, e.g. right click/open in new tab.
```

Comment:
What about navigations that target this window, but originate from another window?

What about navigations that originate from the URL bar or back/forward buttons?

Reply:
Yeah, that's the big TODO.

> _TODO: how do we avoid trapping people? Probably they are only cancelable for same-document navigations? Maybe only for same-origin navigations?_
>
> _Need a more comprehensive "threat model" and goal model for intercepting navigations. Intercepting cross-origin navs might be useful in some cases... and is already possible with `<a>` rewriting. But e.g. back button interception is not allowed._